### PR TITLE
Make the function token the default to run function

### PR DESCRIPTION
**Pull request by yanndupis**
[CAPE-954](https://capeprivacy.atlassian.net/browse/CAPE-954)

This PR add the following:
- `cape.run` will only run if a function token is provided the `FunctionRef`. The user won't be able to call a function just using a user access token. 
- Add a `from_json` method  to `FunctionRef` to load json file generated by the cli with `cape token <FUNCTION_ID> --function-checksum <FUNCTION_CHECKSUM> -o json`
- Update `cli.deploy` and `cli.token` to reflect the latest change made to the cli and generate function token as default
- Update examples
- Update Readme

### CODE REVIEW

File: examples/mean/mean_token.json [hunk 1]
```diff
@@ -0,0 +1 @@
+~/Documents/cape/pycape ~/Documents/cape/pycape/examples ~/Documents/cape/pycape/examples/echo
```
Comment (jvmncs) [inline examples/mean/mean_token.json]: Mistake?

File: pycape/cape.py [hunk 2]
```diff
@@ -14,7 +14,8 @@
 
     cape = Cape()
 
-    cape.connect("9712r5dynf57l1rcns2")
+    f = FunctionRef.from_json("function_token.json")
+    cape.connect()
```
Comment (jvmncs) [inline pycape/cape.py]: Missing a parameter here

File: pycape/experimental/cli.py [hunk 3]
```diff
@@ -90,47 +95,48 @@ async def token(function_id, expires=None, url: Optional[str] = None):
...
-    url = url or cape_config.ENCLAVE_HOST
-
-    if expires:
-        cmd_token = f"cape token {function_id} --expires {expires} -u {url}"
+    if expiry:
+        cmd_token = (
+            f"cape token {function_id} --function-checksum {function_checksum} "
+            f"--expiry {expiry} -o json"
+        )
+        # breakpoint()
```
Comment (jvmncs) [inline pycape/experimental/cli.py]: Leftover?

File: pycape/cape.py [hunk 4]
```diff
@@ -585,18 +577,25 @@ def _handle_expected_field(dictionary, field, *, fallback_err=None):
...
+    if isinstance(function_ref, fref.FunctionRef):
+        if function_ref.id is None:
+            raise ValueError(
+                "The function ID in the function reference " "provided is empty."
+            )
+
+        if function_ref.token is None:
+            raise ValueError(
+                "The function token in the function reference " "provided is empty."
+            )
```
Comment (jvmncs) [inline pycape/cape.py]: any reason these strings are broken?

File: pycape/cape.py [hunk 5]
```diff
@@ -585,18 +577,25 @@ def _handle_expected_field(dictionary, field, *, fallback_err=None):
...
+
         return function_ref
     else:
         raise TypeError(
-            "`function_ref` arg must be a string function ID, "
-            f"or a FunctionRef object, found {type(function_ref)}"
+            f"`function_ref` a FunctionRef object found {type(function_ref)}."
```
Comment (jvmncs) [inline pycape/cape.py]: think this error message needs to be completed

File: pycape/cape.py [hunk 6]
```diff
@@ -313,8 +309,8 @@ async def run(
         preferred when the caller doesn't need to invoke a Cape function more than once.
 
         Args:
-            function_ref: A function ID string or :class:`~.function_ref.FunctionRef`
-                representing a deployed Cape function.
+            function_ref: A function :class:`~.function_ref.FunctionRef` representing
```
Comment (jvmncs) [inline pycape/cape.py]:
```suggestion
            function_ref: A :class:`~.function_ref.FunctionRef` representing
```

**Review by jvmncs — COMMENTED** [review]
a couple other issues:
- all `.DS_Store` files need to be removed. i think we can put this in the gitignore if that's more convenient
- if we are removing the `access_token`, we should also remove the `FunctionAuthType` type in `function_ref.py`. the only reason that exists is to specify which auth subprotocol should be used in websocket messages relating to a particular `FunctionRef`, but without `access_token` there is only one type of auth

File: pycape/cape.py (hunk 2 above)
Comment (jvmncs) [reply pycape/cape.py]:
```suggestion
    cape.connect(f)
```

File: pycape/cape.py (hunk 4 above)
Comment (jvmncs) [reply pycape/cape.py]:
```suggestion
        if function_ref.id is None:
            raise ValueError(
                "The function ID in the provided FunctionRef is empty."
            )

        if function_ref.token is None:
            raise ValueError(
                "The function token in the provided FunctionRef is empty."
            )
```

File: pycape/experimental/cli.py (hunk 3 above)
Comment (jvmncs) [reply pycape/experimental/cli.py]:
```suggestion
```

**Review by jvmncs — CHANGES_REQUESTED** [review]
after talking a bit w/ @gavinuhma, we decided we should keep the `access_token` auth pattern in pycape, at least for now. the main reason is that `Cape.key` / `Cape.encrypt` don't yet work when using function tokens. I think the other changes are good, so the PR is still worth merging

**Comment by yanndupis** [comment]
> after talking a bit w/ @gavinuhma, we decided we should keep the access_token auth pattern in pycape, at least for now. the main reason is that Cape.key / Cape.encrypt don't yet work when using function tokens. I think the other changes are good, so the PR is still worth merging

ok @jvmncs @gavinuhma  but if we keep access token what do we tell user to run a function? for example in all the examples and usage? to use access token or function token? 

this?
```
from pycape import Cape
from pycape import FunctionRef

client = Cape(url="wss://enclave.capeprivacy.com")
function_id = "2heeV48kLCAsvj6nYY87Fh"
function_checksum = "cbca8c9f7ac41138935018c3f45cd16d1abfbe15a37b1fc09a11dfbc3d44b447"
f = FunctionRef(function_id, function_checksum)
result = client.run(f, b"Hello!")
print(result.decode())
# Hello!
```
or this?
```
from pycape import Cape
from pycape import FunctionRef

client = Cape(url="wss://enclave.capeprivacy.com")
f = FunctionRef.from_json("echo_token.json")
result = client.run(f, b"Hello!")
print(result.decode())
# Hello!
```

File: pycape/cape.py (hunk 4 above)
Comment (yanndupis) [reply pycape/cape.py]: oops thanks

File: pycape/function_ref.py [hunk 7]
```diff
@@ -46,32 +28,34 @@ class FunctionRef:
     Args:
         id: Required string denoting the function ID of the deployed Cape function.
             Typically given with the output of the Cape CLI's ``deploy`` command.
+        token: Required string containing a Cape function token generated by the Cape
+            CLI during ``cape token``. If ``None``, the Cape access token provided to
+            :class:`~pycape.Cape` will be used by :meth:`.Cape.connect` /
+            :meth:`.Cape.run` instead.
```
Comment (jvmncs) [inline pycape/function_ref.py]:
```suggestion
            CLI during ``cape token``.
```

File: pycape/cape.py [hunk 8]
```diff
@@ -14,7 +14,8 @@
 
     cape = Cape()
 
-    cape.connect("9712r5dynf57l1rcns2")
+    f = FunctionRef.from_json("function_token.json")
```
Comment (jvmncs) [inline pycape/cape.py]:
```suggestion
    f = FunctionRef.from_json("function.json")
```

File: pycape/cape.py [hunk 9]
```diff
@@ -188,8 +183,9 @@ async def function_context(
         **Usage** ::
 
             cape = Cape(url="wss://enclave.capeprivacy.com")
+            f = FunctionRef.from_json("function_token.json")
```
Comment (jvmncs) [inline pycape/cape.py]:
```suggestion
            f = FunctionRef.from_json("function.json")
```

File: pycape/experimental/cli.py [hunk 10]
```diff
@@ -90,47 +95,47 @@ async def token(function_id, expires=None, url: Optional[str] = None):
 
     Args:
         function_id: A function ID string representing a deployed Cape function.
-        exprires: Amount of time in seconds until the the function token expires.
-    url: The Cape platform's websocket URL, which is responsible for forwarding
-            client requests to the proper enclave instances. If None, tries to load
-            value from the ``CAPE_ENCLAVE_HOST`` environment variable. If no such
-            variable value is supplied, defaults to ``"wss://enclave.capeprivacy.com"``.
+        expriry: Amount of time in seconds until the function token expires.
```
Comment (jvmncs) [inline pycape/experimental/cli.py]:
```suggestion
        expiry: Amount of time in seconds until the function token expires.
```

File: pycape/function_ref.py [hunk 11]
```diff
@@ -85,15 +69,45 @@ def checksum(self):
     def token(self):
         return self._token
 
-    @property
-    def auth_type(self):
-        return self._auth_type
+    @classmethod
+    def from_json(cls, token_path: Union[str, os.PathLike]):
```
Comment (jvmncs) [inline pycape/function_ref.py:71]: `to_json` would be good to have as well I think, this can be a TODO

File: examples/run_echo.py [hunk 12]
```diff
@@ -1,20 +1,16 @@
...
 from pycape import Cape
 from pycape import FunctionRef
 
 if __name__ == "__main__":
-    token = os.environ.get("CAPE_TOKEN", None)
     url = os.environ.get("CAPE_HOST", "wss://enclave.capeprivacy.com")
-    function_id = os.environ.get("CAPE_FUNCTION_ID", "VNgMtygWv8wCwwjbbQ2kH6")
-    function_checksum = os.environ.get("CAPE_FUNCTION_CHECKSUM", None)
+    token_file = os.environ.get("TOKEN_FILE", "echo_token.json")
+    token_file = pathlib.Path(__file__).parent.absolute() / token_file
```
Comment (jvmncs) [inline examples/run_echo.py]:
```suggestion
    function_json = os.environ.get("FUNCTION_JSON", "echo_token.json")
    function_json = pathlib.Path(__file__).parent.absolute() / function_json
```

File: examples/run_echo.py [hunk 13]
```diff
@@ -1,20 +1,16 @@
...
 
-    if function_checksum is None:
-        function_ref = function_id
-    else:
-        function_ref = FunctionRef(function_id, function_checksum)
-
-    cape = Cape(url=url, access_token=token)
+    function_ref = FunctionRef.from_json(token_file)
```
Comment (jvmncs) [inline examples/run_echo.py]:
```suggestion
    function_ref = FunctionRef.from_json(function_json)
```

File: examples/run_echo.py (hunk 12 above)
Comment (jvmncs) [reply examples/run_echo.py]: this renaming should be same for the other examples that use this pattern

**Review by jvmncs — APPROVED** [review]
lgtm w/ comments. the main issue i see is that we are referring to the function JSONs as "token files" when that is not a good name, since these files contain more than just the token

File: pycape/function_ref.py (hunk 11 above)
Comment (yanndupis) [reply pycape/function_ref.py:71]: good idea!

Comment (yanndupis) [reply pycape/function_ref.py:71]: https://capeprivacy.atlassian.net/browse/CAPE-1048

**Pull request merged**